galaxy/api/base.py

```diff
@@ -0,0 +1,204 @@
+# (c) 2012-2019, Ansible by Red Hat
+#
+# This file is part of Ansible Galaxy
+#
+# Ansible Galaxy is free software: you can redistribute it and/or modify
+# it under the terms of the Apache License as published by
+# the Apache Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# Ansible Galaxy is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# Apache License for more details.
+#
+# You should have received a copy of the Apache License
+# along with Galaxy.  If not, see <http://www.apache.org/licenses/>.
+from django.core.exceptions import PermissionDenied
+from django.http import Http404
+
+from rest_framework import exceptions
+from rest_framework import generics
+from rest_framework import mixins
+from rest_framework.response import Response
+from rest_framework.settings import api_settings
+from rest_framework import status as http_codes
+from rest_framework import views
+
+
+def exception_handler(exc, context):
+    if isinstance(exc, Http404):
+        msg = 'Not found.'
+        data = {'message': str(msg)}
+
+        views.set_rollback()
+        return Response(data, status=http_codes.HTTP_404_NOT_FOUND)
+
+    if isinstance(exc, PermissionDenied):
+        msg = 'Permission denied.'
+        data = {'message': str(msg)}
+
+        views.set_rollback()
+        return Response(data, status=http_codes.HTTP_403_FORBIDDEN)
+
+    if isinstance(exc, exceptions.APIException):
+        headers = {}
+        if getattr(exc, 'auth_header', None):
+            headers['WWW-Authenticate'] = exc.auth_header
+        if getattr(exc, 'wait', None):
+            headers['Retry-After'] = '%d' % exc.wait
+
+        data = {'message': exc.default_detail}
+        errors = []
+
+        if isinstance(exc.detail, list):
+            errors = [{'message': e} for e in exc.detail]
+        elif isinstance(exc.detail, dict):
```

Comment:
Can exc.detail be anything other than a list?  https://www.django-rest-framework.org/api-guide/exceptions/#inspecting-api-exceptions claims:

> .detail - Return the textual description of the error.

The django_rest_framework.exceptions.APIException subclasses all seem to set the `detail` attribute to text.

Reply:
Looks like drf.exceptions._get_exception_details() can return a list, .detail can be other things that text.

Reply:
DRF documentation is incorrect. 

According to implementation of `APIException` class and standard `exception_handler`, the `.detail` attribute may be either `list` or its subclass, `dict` or its subclass and string (specifically ErrorDetail object, which inherits string and has `.code` attribute).